Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.cpp

```diff
@@ -55,161 +55,81 @@ FileSystemSyncAccessHandle::FileSystemSyncAccessHandle(ScriptExecutionContext& c
 }
 
-void FileSystemSyncAccessHandle::closeInternal(CloseCallback&& callback)
+ExceptionOr<void> FileSystemSyncAccessHandle::flush()
 {
-    if (m_closeResult)
-        return callback(ExceptionOr<void> { *m_closeResult });
+    if (m_isClosed)
+        return Exception { InvalidStateError, "AccessHandle is closed"_s };
 
-    auto isClosing = !m_closeCallbacks.isEmpty();
-    m_closeCallbacks.append(WTFMove(callback));
-    if (isClosing)
-        return;
+    auto* scope = downcast<WorkerGlobalScope>(scriptExecutionContext());
+    if (!scope)
+        return Exception { InvalidStateError, "Context is invalid"_s };
 
-    ASSERT(m_file);
-    closeFile();
+    bool succeeded = FileSystem::flushFile(m_file.handle());
+    return succeeded ? ExceptionOr<void> { } : Exception { UnknownError };
 }
 
-void FileSystemSyncAccessHandle::closeFile()
+ExceptionOr<void> FileSystemSyncAccessHandle::close()
 {
-    if (!m_file)
-        return;
+    if (m_isClosed)
+        return Exception { InvalidStateError, "AccessHandle is closed"_s };
 
     auto* scope = downcast<WorkerGlobalScope>(scriptExecutionContext());
-    ASSERT(scope);
-
-    WorkerGlobalScope::postFileSystemStorageTask([weakThis = WeakPtr { *this }, file = std::exchange(m_file, { }), workerThread = Ref { scope->thread() }]() mutable {
-        workerThread->runLoop().postTask([weakThis = WTFMove(weakThis)](auto&) mutable {
-            if (weakThis)
-                weakThis->didCloseFile();
-        });
-    });
-}
+    if (!scope)
+        return Exception { InvalidStateError, "Context is invalid"_s };
 
-void FileSystemSyncAccessHandle::didCloseFile()
-{
-    closeBackend(CloseMode::Async);
+    closeInternal(ShouldNotifyBackend::Yes);
+    return { };
 }
 
-void FileSystemSyncAccessHandle::closeBackend(CloseMode mode)
+void FileSystemSyncAccessHandle::closeInternal(ShouldNotifyBackend shouldNotifyBackend)
 {
-    if (m_closeResult)
-        return;
-
-    if (mode == CloseMode::Async) {
-        m_source->closeSyncAccessHandle(m_identifier, [this, protectedThis = Ref { *this }](auto result) mutable {
-            didCloseBackend(WTFMove(result));
-        });
+    if (m_isClosed)
         return;
-    }
 
-    m_source->closeSyncAccessHandle(m_identifier, [](auto) { });
-    didCloseBackend({ });
-}
-
-void FileSystemSyncAccessHandle::didCloseBackend(ExceptionOr<void>&& result)
-{
-    if (m_closeResult)
-        return;
+    m_isClosed = true;
+    ASSERT(m_file);
+    m_file = { };
 
-    m_closeResult = WTFMove(result);
-    auto callbacks = std::exchange(m_closeCallbacks, { });
-    for (auto& callback : callbacks)
-        callback(ExceptionOr<void> { *m_closeResult });
+    if (shouldNotifyBackend == ShouldNotifyBackend::Yes)
+        m_source->closeSyncAccessHandle(m_identifier);
 }
 
 ExceptionOr<unsigned long long> FileSystemSyncAccessHandle::read(BufferSource&& buffer, FileSystemSyncAccessHandle::FilesystemReadWriteOptions options)
 {
     ASSERT(!isMainThread());
```

Comment:
We can either add these asserts to the other methods or remove these asserts since all methods are doing IO now in the same thread.

Reply:
Sure; will remove.